peps/pep-0667.rst

```diff
@@ -40,72 +63,228 @@ The current implementation of ``locals()`` and ``frame.f_locals`` is slow,
-that is created on the fly from the array of local variables.
+that is created on the fly from the array of local variables. The
+``PyFrame_LocalsToFast()`` C API is then called by debuggers and trace
+functions that want to write their changes back to the array (until
+Python 3.11, this API was called implicitly after every trace function
+invocation rather than being called explicitly by the trace functions).
+
 This can result in the array and dictionary getting out of sync with
-each other. Writes to the ``f_locals`` may not show up as
-modifications to local variables. Writes to local variables can get lost.
+each other. Writes to the ``f_locals`` frame atribute may not show up as
+modifications to local variables if ``PyFrame_LocalsToFast()`` is never
+called. Writes to local variables can get lost if the dictionary snapshot
+is not refreshed before being written back to the frame.
 
 By making ``frame.f_locals`` return a view on the
 underlying frame, these problems go away. ``frame.f_locals`` is always in
 sync with the frame because it is a view of it, not a copy of it.
 
+Making the ``locals()`` builtin return independent snapshots
+------------------------------------------------------------
+
+When ``exec`` was converted from a statement to a builtin function
+in Python 3.0 (part of the core language changes in :pep:`3100`), the
+associated implicit call to ``PyFrame_LocalsToFast()`` was removed, so
+it typically appears as if attempts to write to local variables with
+``exec`` in optimized frames are ignored::
+
+    >>> def f():
+    ...     x = 0
+    ...     exec("x = 1")
+    ...     print(x)
+    ...     print(locals()["x"])
+    ...
+    >>> f()
+    0
+    0
+
+In truth, the writes aren't being ignored, they just aren't
+being copied from the dictionary cache back to the optimized local
+variable array. The changes to the dictonary are then overwritten
+the next time the dictionary cache is refreshed from the array::
+
+    >>> def f():
+    ...     x = 0
+    ...     locals_cache = locals()
+    ...     exec("x = 1")
+    ...     print(x)
+    ...     print(locals_cache["x"])
+    ...     print(locals()["x"])
+    ...
+    >>> f()
+    0
+    1
+    0
+
+The behaviour becomes even stranger if a tracing function
+or another piece of code invokes ``PyLocals_ToFast()`` before
+the cache is next refreshed, as in those cases the change *is*
+written back to the optimized local variable array::
+
+    >>> from sys import _getframe
+    >>> from ctypes import pythonapi, py_object, c_int
+    >>> _locals_to_fast = pythonapi.PyFrame_LocalsToFast
+    >>> _locals_to_fast.argtypes = [py_object, c_int]
+    >>> def f():
+    ...     _frame = _getframe()
+    ...     _f_locals = _frame.f_locals
+    ...     x = 0
+    ...     exec("x = 1")
+    ...     _locals_to_fast(_frame, 0)
+    ...     print(x)
+    ...     print(locals()["x"])
+    ...     print(_f_locals["x"])
+    ...
+    >>> f()
+    1
+    1
+    1
```

Comment:
I didn't want to get into it in the PEP itself, but here's a fun headscratcher for you: in Python 3.12, if you delay retrieving `_frame.f_locals` until after the exec+PyFrame_LocalsToFast calls, the local variables don't get updated. If you retrieve it beforehand, they get updated, even if you immediately drop the reference and get a new one afterwards.

Late `f_locals` retrieval:

```
>>> def f():
...     _frame = _getframe()
...     x = 0
...     print(locals()["x"])
...     exec("x = 1")
...     _locals_to_fast(_frame, 0)
...     print(x)
...     print(locals()["x"])
...     print(_frame.f_locals["x"])
...
>>> f()
0
0
0
0
```

Early `f_locals` retrieval:

```
>>> def f():
...     _frame = _getframe()
...     x = 0
...     print(_frame.f_locals["x"])
...     exec("x = 1")
...     _locals_to_fast(_frame, 0)
...     print(x)
...     print(locals()["x"])
...     print(_frame.f_locals["x"])
...
>>> f()
0
1
1
1
>>>
```

I'm genuinely not sure what is going on there, since `PyEval_GetLocals()` should be implicitly populating `frame.f_locals` so it has a borrowed reference to return.